### ISSUE TITLE
preCICE stops coupling before finishing the last time window if `max-time` is no integer multiple of `time-window-size`

### ISSUE THREAD

**Issue by NiklasVin** [precice/precice#1922]
**Describe your setup**

Operating system: Ubuntu 22.04.3 
preCICE Version: Commit a174c66e669d4e0e700679185a577439833e6c4e

**Describe the problem**
If `max-time` in the preCICE config file is not an integer multiple of `time-window-size`, 
the minimal number of iterations is not executed, and `getMaxTimeStepSize()` does not return a value that the coupling ends exactly at `max-time`.
<details>
<summary>The output of the last few iterations from the participant SolverOne </summary>

```console
preCICE: iteration: 1 of 2 (min 2), time-window: 2, time: 0.75 of 2, time-window-size: 0.75, max-time-step-size: 0.75, ongoing: yes, time-window-complete: yes, write-iteration-checkpoint 
DUMMY: Advancing in time
DUMMY: Writing iteration checkpoint
preCICE: iteration: 1 of 2 (min 2), time-window: 2, time: 1.01 of 2, time-window-size: 0.75, max-time-step-size: 0.49, ongoing: yes, time-window-complete: no, 
DUMMY: Advancing in time
preCICE: iteration: 1 of 2 (min 2), time-window: 2, time: 1.27 of 2, time-window-size: 0.75, max-time-step-size: 0.23, ongoing: yes, time-window-complete: no, 
DUMMY: Advancing in time
preCICE: iteration: 2 of 2 (min 2), time-window: 2, time: 0.75 of 2, time-window-size: 0.75, max-time-step-size: 0.75, ongoing: yes, time-window-complete: no, read-iteration-checkpoint 
DUMMY: Reading iteration checkpoint
preCICE: iteration: 2 of 2 (min 2), time-window: 2, time: 1.01 of 2, time-window-size: 0.75, max-time-step-size: 0.49, ongoing: yes, time-window-complete: no, 
DUMMY: Advancing in time
preCICE: iteration: 2 of 2 (min 2), time-window: 2, time: 1.27 of 2, time-window-size: 0.75, max-time-step-size: 0.23, ongoing: yes, time-window-complete: no, 
DUMMY: Advancing in time
preCICE: Time window completed
preCICE: iteration: 1 of 2 (min 2), time-window: 3, time: 1.5 of 2, time-window-size: 0.75, max-time-step-size: 0.75, ongoing: yes, time-window-complete: yes, write-iteration-checkpoint 
DUMMY: Advancing in time
DUMMY: Writing iteration checkpoint
preCICE: iteration: 1 of 2 (min 2), time-window: 3, time: 1.76 of 2, time-window-size: 0.75, max-time-step-size: 0.49, ongoing: yes, time-window-complete: no, 
DUMMY: Advancing in time
preCICE: iteration: 1 of 2 (min 2), time-window: 3, time: 2.02 of 2, time-window-size: 0.75, max-time-step-size: 0, ongoing: no, time-window-complete: no, 
DUMMY: Advancing in time
preCICE: Close communication channels
DUMMY: Closing C++ solver dummy...
```

</details>

<details>
<summary>The output of the last few iterations from the participant SolverTwo</summary>

```console
preCICE: iteration: 1 of 2 (min 2), time-window: 2, time: 0.75 of 2, time-window-size: 0.75, max-time-step-size: 0.75, ongoing: yes, time-window-complete: yes, write-iteration-checkpoint 
DUMMY: Advancing in time
DUMMY: Writing iteration checkpoint
preCICE: iteration: 1 of 2 (min 2), time-window: 2, time: 1.01 of 2, time-window-size: 0.75, max-time-step-size: 0.49, ongoing: yes, time-window-complete: no, 
DUMMY: Advancing in time
preCICE: iteration: 1 of 2 (min 2), time-window: 2, time: 1.27 of 2, time-window-size: 0.75, max-time-step-size: 0.23, ongoing: yes, time-window-complete: no, 
DUMMY: Advancing in time
preCICE: Mapping "Data-Two" for t=0.75 from "SolverTwo-Mesh" to "SolverOne-Mesh"
preCICE: Mapping "Data-Two" for t=1.01 from "SolverTwo-Mesh" to "SolverOne-Mesh"
preCICE: Mapping "Data-Two" for t=1.27 from "SolverTwo-Mesh" to "SolverOne-Mesh"
preCICE: Mapping "Data-Two" for t=1.5 from "SolverTwo-Mesh" to "SolverOne-Mesh"
preCICE: No converge measures defined.
preCICE: Mapping "Data-One" for t=0.75 from "SolverOne-Mesh" to "SolverTwo-Mesh"
preCICE: Mapping "Data-One" for t=1.5 from "SolverOne-Mesh" to "SolverTwo-Mesh"
preCICE: iteration: 2 of 2 (min 2), time-window: 2, time: 0.75 of 2, time-window-size: 0.75, max-time-step-size: 0.75, ongoing: yes, time-window-complete: no, read-iteration-checkpoint 
DUMMY: Reading iteration checkpoint
preCICE: iteration: 2 of 2 (min 2), time-window: 2, time: 1.01 of 2, time-window-size: 0.75, max-time-step-size: 0.49, ongoing: yes, time-window-complete: no, 
DUMMY: Advancing in time
preCICE: iteration: 2 of 2 (min 2), time-window: 2, time: 1.27 of 2, time-window-size: 0.75, max-time-step-size: 0.23, ongoing: yes, time-window-complete: no, 
DUMMY: Advancing in time
preCICE: Mapping "Data-Two" for t=0.75 from "SolverTwo-Mesh" to "SolverOne-Mesh"
preCICE: Mapping "Data-Two" for t=1.01 from "SolverTwo-Mesh" to "SolverOne-Mesh"
preCICE: Mapping "Data-Two" for t=1.27 from "SolverTwo-Mesh" to "SolverOne-Mesh"
preCICE: Mapping "Data-Two" for t=1.5 from "SolverTwo-Mesh" to "SolverOne-Mesh"
preCICE: No converge measures defined.
preCICE: Time window completed
preCICE: Mapping "Data-One" for t=1.5 from "SolverOne-Mesh" to "SolverTwo-Mesh"
preCICE: Mapping "Data-One" for t=2.25 from "SolverOne-Mesh" to "SolverTwo-Mesh"
preCICE: iteration: 1 of 2 (min 2), time-window: 3, time: 1.5 of 2, time-window-size: 0.75, max-time-step-size: 0.75, ongoing: yes, time-window-complete: yes, write-iteration-checkpoint 
DUMMY: Advancing in time
DUMMY: Writing iteration checkpoint
preCICE: iteration: 1 of 2 (min 2), time-window: 3, time: 1.76 of 2, time-window-size: 0.75, max-time-step-size: 0.49, ongoing: yes, time-window-complete: no, 
DUMMY: Advancing in time
preCICE: iteration: 1 of 2 (min 2), time-window: 3, time: 2.02 of 2, time-window-size: 0.75, max-time-step-size: 0, ongoing: no, time-window-complete: no, 
DUMMY: Advancing in time
preCICE: Close communication channels
DUMMY: Closing C++ solver dummy...
```

</details>

**Step To Reproduce**
1. Edit the examples/solverdummies/cpp/solverdummy.cpp file [here](https://github.com/precice/precice/blob/76f871124a813c59394e4fdcc1da990dce8064ea/examples/solverdummies/cpp/solverdummy.cpp#L72C9-L72C9):
```diff
while (participant.isCouplingOngoing()) {

    if (participant.requiresWritingCheckpoint()) {
      std::cout << "DUMMY: Writing iteration checkpoint\n";
    }

-    double dt = participant.getMaxTimeStepSize();
+    double dt = std::min(0.26, participant.getMaxTimeStepSize());
    participant.readData(meshName, dataReadName, vertexIDs, dt, readData);

    for (int i = 0; i < numberOfVertices * dimensions; i++) {
      writeData.at(i) = readData.at(i) + 1;
    }

```
2. Change the preCICE config file examples/solverdummies/precice-config.xml
```diff
while (participant.isCouplingOngoing()) {
   <coupling-scheme:serial-implicit>
     <participants first="SolverOne" second="SolverTwo" />
-    <max-time-windows value="2" />
-    <time-window-size value="1.0" />
+    <max-time value="2" />
+    <time-window-size value="0.75" />
     <min-iterations value="2" />
     <max-iterations value="2" />
     <exchange data="Data-One" mesh="SolverOne-Mesh" from="SolverOne" to="SolverTwo" />

```
3. Compile the solverdummy.cpp file and run it as SolverOne and SolverTwo with the precice-config.xml file.

**Expected behaviour**
I am not sure how preCICE should behave but there are some behaviours I would expect:
1. `getMaxTimeStepSize()` returns in cases like above in the third time window, not 0.49 but 0.24 if the current time is 1.76 and the `max-time` is 2. So the minimum of the remaining time window time and the remaining coupling time.
2. Alternatively to 1.: Disregarding the `max-time`, preCICE finishes the started time window.
3. Even though the `max-time` is reached or exceeded, preCICE does its minimal coupling iterations.

**Additional context**
The example above is chosen because the coupling ends at a strange time and thus highlights the odd behaviour well.
You observe similar problems if the solverdummy.cpp stays unchanged or if the file is changed as follows:
```diff
-    double dt = participant.getMaxTimeStepSize();
+    double dt = 0.25;
```

**Comment by uekerman** [comment]
No showstopper in my opinion. Rather an edge case, but indeed buggy, yes. And probably never worked.

**Comment by fsimonis** [comment]
This was fixed in #1933.

We now limit the result of `getMaxTimeStepSize()` by `max-time` if its given.
